### PR TITLE
Limit blob file read IO with rate_limiter

### PR DESCRIPTION
**Pull request by wujy-cs**
Limit Titan blob file read IO with rate_limiter, to avoid burst of IO starve online workloads.

### CODE REVIEW

**Comment by wujy-cs** [comment]
@yiwu-arbug Maybe we should use another limiter for reader?

File: src/blob_file_test.cc [hunk 1]
```diff
@@ -88,6 +88,10 @@ class BlobFileTest : public testing::Test {
     const int n = 100;
     std::vector<BlobHandle> handles(n);
 
+    auto rate_limiter =
```
Comment (yiwu-arbug) [inline src/blob_file_test.cc]: For test, add a new test case? The test should create a mock test rate limiter, ran GC job, and check if it call into the mock rate limiter. It should test with all three values of `RateLimiter::Mode`.

File: src/blob_file_cache.cc [hunk 2]
```diff
@@ -73,7 +73,9 @@ Status BlobFileCache::FindFile(uint64_t file_number, uint64_t file_size,
     if (db_options_.advise_random_on_open) {
       f->Hint(RandomAccessFile::RANDOM);
     }
-    file.reset(new RandomAccessFileReader(std::move(f), file_name));
+    file.reset(new RandomAccessFileReader(std::move(f), file_name, nullptr,
+                                          nullptr, 0, nullptr,
+                                          env_options_.rate_limiter, true));
```
Comment (yiwu-arbug) [inline src/blob_file_cache.cc]: To be consistent with rocksdb, we should rate limit GC only, not user reads. Seems `BlobGCJob` currently call `NewBlobFileReader` directly, which does not go through `BlobFileCache`. Let's revert the change here.

**Comment by yiwu-arbug** [comment]
> @yiwu-arbug Maybe we should use another limiter for reader?

Not really. Check `RateLimiter::Mode`.

**Comment by wujy-cs** [comment]
@thanks!

**Comment by wujy-cs** [comment]
@yiwu-arbug Thanks! Please take a look again?

File: src/blob_gc_job_test.cc [hunk 3]
```diff
@@ -260,6 +344,8 @@ TEST_F(BlobGCJobTest, DiscardEntry) { TestDiscardEntry(); }
 
 TEST_F(BlobGCJobTest, RunGC) { TestRunGC(); }
 
+TEST_F(BlobGCJobTest, GCLimiter) { TestGCLimiter(); }
```
Comment (yiwu-arbug) [inline src/blob_gc_job_test.cc]: I think we don't need to have a separate `TestGCLimiter()`. Just put the test code inside `BlobGCJobTest` body.

File: src/blob_gc_job_test.cc [hunk 4]
```diff
@@ -196,6 +210,76 @@ class BlobGCJobTest : public testing::Test {
     DestroyDB();
   }
 
+  void TestGCLimiter() {
+    class TestLimiter : public RateLimiter {
+     public:
+      TestLimiter(RateLimiter::Mode mode)
+          : RateLimiter(mode), read(false), write(false) {}
+
+      virtual size_t RequestToken(size_t bytes, size_t alignment,
```
Comment (yiwu-arbug) [inline src/blob_gc_job_test.cc]: remove `virtual` keyword and add `override` keyword. Same for other methods.

File: src/blob_gc_job_test.cc [hunk 5]
```diff
@@ -196,6 +210,76 @@ class BlobGCJobTest : public testing::Test {
...
+      }
+
+      virtual int64_t GetBytesPerSecond() const { return 0; }
+
+      bool ReadRequested() { return read; }
+
+      bool WriteRequested() { return write; }
+
+     private:
+      bool read;
```
Comment (yiwu-arbug) [inline src/blob_gc_job_test.cc]: need to reset both of these flags to false after flush, since flush also make use of rate limiter.

File: src/blob_gc_job_test.cc [hunk 6]
```diff
@@ -260,6 +261,98 @@ TEST_F(BlobGCJobTest, DiscardEntry) { TestDiscardEntry(); }
...
+    bool ReadRequested() { return read; }
+
+    bool WriteRequested() { return write; }
+
+   private:
+    bool read;
+    bool write;
+  };
+
+  auto TriggerGC = [this] {
```
Comment (yiwu-arbug) [inline src/blob_gc_job_test.cc]: give it a better name? It's not triggering GC now.
